src/GafferScene/FramingConstraint.cpp

```diff
@@ -0,0 +1,687 @@
+			// Dot each corner with each plane normal and compare to the appropriate result
+			float dot = boxSpaceNormals[i].dot( corner );
+			if( minDists )
+			{
+				(*minDists)[i] = std::min( (*minDists)[i], dot );
+			}
+
+			if( maxDists )
+			{
+				(*maxDists)[i] = std::max( (*maxDists)[i], dot );
+			}
+		}
+	}
+
+	// Add an offset to all results to account for the transform of the boxes origin
+	V3f boxOriginWorld = boxTransform.translation();
+	for( unsigned int i = 0; i < planeNormals.size(); i++ )
+	{
+		float dot = planeNormals[i].dot( boxOriginWorld - planeOrigin );
+		if( minDists )
+		{
+			(*minDists)[i] += dot;
+		}
+
+		if( maxDists )
+		{
+			(*maxDists)[i] += dot;
+		}
+	}
+}
+
+// The same signature as planeDistsFromBox, except for a sphere specified by a center point and a
+// a point on the surface of the sphere
+inline void planeDistsFromSphere(
+	const V3f &center, const V3f &pointOnSurface, const M44f &sphereTransform,
+	const V3f &planeOrigin, std::vector< V3f > &planeNormals,
+	std::vector<float> *minDists, std::vector<float> *maxDists
+)
+{
+	V3f worldCenter = center * sphereTransform;
+	float radius = ( pointOnSurface * sphereTransform - worldCenter ).length();
+
+	V3f disp = worldCenter - planeOrigin;
+
+	if( minDists )
+	{
+		minDists->resize( planeNormals.size() );
+	}
+
+	if( maxDists )
+	{
+		maxDists->resize( planeNormals.size() );
+	}
+
+	for( unsigned int i = 0; i < planeNormals.size(); i++ )
+	{
+		float dot = planeNormals[i].dot( disp );
+		float scaledRad = radius * planeNormals[i].length();
+		if( minDists )
+		{
+			(*minDists)[i] = dot - scaledRad;
+		}
+
+		if( maxDists )
+		{
+			(*maxDists)[i] = dot + scaledRad;
+		}
+	}
+}
+
+// Set an ObjectVector plug to store a transform and object
+void setTransformAndObject( Gaffer::ValuePlug *output, const M44f &transform, const IECore::Object *object )
+{
+	ObjectVectorPtr result = new ObjectVector();
+	result->members().resize( 2 );
+	result->members()[0] = new M44fData( transform );
+	// const_cast is safe because the resulting ObjectVector as a whole is treated as const
+	result->members()[1] = const_cast< IECore::Object *>( object );
+
+	static_cast<ObjectVectorPlug *>( output )->setValue( result );
+}
+
+} // namespace
+
+void FramingConstraint::hash( const Gaffer::ValuePlug *output, const Gaffer::Context *context, IECore::MurmurHash &h ) const
+{
+	SceneElementProcessor::hash( output, context, h );
+	if( output != transformAndObjectPlug() )
+	{
+		return;
+	}
+
+	const ScenePath &path = context->get<ScenePath>( ScenePlug::scenePathContextName );
+	const ScenePath parentPath( path.begin(), path.end() - 1 );
+	inPlug()->transformPlug()->hash( h );
+	h.append( inPlug()->fullTransformHash( parentPath ) );
+	inPlug()->objectPlug()->hash( h );
+
+	auto targetOpt = target();
+	if( !targetOpt )
+	{
+		return;
+	}
+
+	boundModePlug()->hash( h );
+	extendFarClipPlug()->hash( h );
+	paddingPlug()->hash( h );
+
+	std::unique_ptr<Context::EditableScope> useFrameScope;
```

Comment:
This requires a heap allocation in `make_unique` - `std::optional` would provide the same functionality but allocate in place.